Add getAccount and enhance account filtering (#6499)
AzureAD/microsoft-authentication-library-for-js: lib/msal-browser/docs/accounts.md

The `@azure/msal-browser` library provides the following APIs to access cached accounts:

-   `getAllAccounts()`: returns all the accounts currently in the cache. Supports an optional filter to return a specific set of accounts. An application must choose an account to acquire tokens silently.
-   `getAccount()`: returns the first cached account that matches the filter passed in. The order in which accounts are read from the cache is arbitrary and there is no guarantee that the first account in the filtered list will be the same for any two calls of `getAccount`. As explained below, increasing the number of filter attributes will provide more exact matches.

### Account Filter Object

> The [AccountFilter](https://azuread.github.io/microsoft-authentication-library-for-js/ref/types/_azure_msal_common.AccountFilter.html) type documentation lists the properties that can be used and combined to filter accounts.

Author: This link's destination doesn't exist yet because the type wasn't public before. If the link doesn't work I'll replace it once type docs are published.

> Note: A single account filter attribute is usually not guaranteed to uniquely identify a cached account object. Adding a combination of attributes that don't repeat together, such as `homeAccountId` + `localAccountId`, can help refine the search.

> > Note: `realm` is the name given to `tenantId` in the cache.

Reviewer: ```suggestion
> Note: `realm` is `tenantId` in the cache.
```

The `@azure/msal-browser` library provides the following APIs to access cached accounts:

> -   `getAllAccounts()`: returns all the accounts currently in the cache. A filter can be passed in to narrow down the returned accounts. An application must choose an account to acquire tokens silently.

Reviewer: ```suggestion
-   `getAllAccounts()`: returns all the accounts currently in the cache. Supports a filter param to return a specific set of accounts. An application must choose an account to acquire tokens silently.
```

> -   `getAccountByFilter()`: returns the first cached account that matches the filter passed in. The order in which accounts are read from the cache is arbitrary and there is no guarantee that the first account in the filtered list will be the same for any two calls of `getAccountByFilter`.

Reviewer: ```suggestion
-   `getAccountByFilter()`: returns the first cached account that matches the filter passed in. The order in which accounts are read from the cache is arbitrary and there is no guarantee that the first account in the filtered list will be the same for any two calls of `getAccountByFilter`. Please check below for guidance to make a unique match possible.
```

### Account Filter Object

The following table presents the properties that can be used in the `AccountFilter` object that can be passed into the APIs listed above.

> Note: A single account filter attribute is usually not guaranteed to uniquely identify a cached account object. Adding a combination of attributes that don't repeat together, such as `homeAccountId` + `localAccountId`, can help refine the search.

> | Property Name        | Description                                                                                                                                                                                  | Example                                                                   |

Reviewer 2: Let's reference the typedocs here instead of writing it all out - less to maintain later :)

Author: I knew this was coming 😆

| `username`           | The `preferred_username` claim of the id_token that represents this account                                                                                                                  | test@example.com                                                          |
| `localAccountId`     | Local, tenant-specific account identifer for this account object, usually used in legacy cases                                                                                               | 00000000-0000-0000-0000-000000000000                                      |
| `name`               | Full name for the account, including given name and family name.                                                                                                                             | John Doe                                                                  |
| `nativeAccountId`    | The user's native account ID                                                                                                                                                                 | 00000000-0000-0000-0000-000000000000                                      |

> > Note: `realm` is the name that `tenantId` receives in the cache.

Reviewer: nit: 
```suggestion
> Note: `realm` is the name that `tenantId` received in the cache.
```

Author: Changed the wording

> The following `getAccountBy` APIs are marked for deprecation and will be removed in a future version of MSAL. Please migrate to `getAccountByFilter()`:

Reviewer: I assume you added `@deprecated` tag for typedocs too.

Author: Yep you can verify that in the diff for PublicClientApplication.

});
```

#### Using `username``

> Note: The `username` value can included in the `AccountFilter` object as either `username` or `loginHint`. This is because the `username` claim is one of the 3 values (along with the `login_hint` and `upn` ID token claims) that the token service accepts as login hint. If your application is certain that the value in question is a `username`, setting it as the `AccountFilter.username` property will yield better search performance. Being able to set a `username` value as `loginHint` is useful if your application utilizes a login hint and does not keep context on whether that value came from a `username`, `login_hint`, or `upn` claim.

Passing `username` as `loginHint`

```javascript
const accountUsername = userProfile.username;
const accountFilter = {
    loginHint: accountUsername;
};
request.account = myMSALObj.getAccount(accountFilter);
return await myMSALObj.acquireTokenSilent(request).catch(async (error) => {
    // Handle error
    return await myMSALObj.acquireTokenPopup(request);
});
```

Passing `username` as `username`

```javascript
const accountUsername = userProfile.username;
const accountFilter = {

>     username: accountUsername; // More performant filtering

Reviewer 2: nit: The explanation above is sufficient, I don't want to inadvertently encourage folks to use username instead of the claim
```suggestion
    username: accountUsername;
```

};
request.account = myMSALObj.getAccount(accountFilter);
return await myMSALObj.acquireTokenSilent(request).catch(async (error) => {
    // Handle error
    return await myMSALObj.acquireTokenPopup(request);
});
```

#### Using `upn` claim

```javascript
const accountFilter = {
    loginHint: previouslyObtainedIdTokenClaims.upn;
};
request.account = myMSALObj.getAccount(accountFilter);
return await myMSALObj.acquireTokenSilent(request).catch(async (error) => {
    // Handle error
    return await myMSALObj.acquireTokenPopup(request);
});
```

## Active Account APIs

The `@azure/msal-browser` library also provides 2 convenient APIs to help you keep track of which account is currently "active" and should be used for token requests.

> -   `getActiveAccount()`: Returns the currently active account

Reviewer: ```suggestion
-   `getActiveAccount()`: Returns the current active account
```

}
```

Now, account properties such as: `homeAccountId`, `localAccountId`, and `username` can be used to look up the cached account before acquiring a token silently:

```javascript
// This method attempts silent token acquisition and falls back on acquireTokenPopup
async function getTokenPopup(request, homeAccountId) {
    // In this case, accounts are filtered by homeAccountId, but more attributes can be added to refine the search and increase the precision of the account filter
    const accountFilter = {
        homeAccountId: homeAccountId,
    };
    request.account = myMSALObj.getAccountByFilter(accountFilter);
    return await myMSALObj.acquireTokenSilent(request).catch(async (error) => {
        // Handle error
        return await myMSALObj.acquireTokenPopup(request);
    });
}
```

### Filtering by login hint

> As of `@azure/msal-browser@3.11.0`, all login hint values can be used to search for and filter accounts. In order to filter by login hint, MSAL will compare the `loginHint` value in the `AccountFilter` object against the following account attributes (in order of precedence) to search for matches:

Reviewer 2: 3.11?

Author: The next release, this will be a minor change so the 3.11 is going to be the next version.

Reviewer 2: We're on 3.1 right now, next release will be 3.2

Author: Sorry my bad, misread :S

-   `login_hint` ID token claim
-   `username` account property
-   `upn` ID token claim

> Note: Regardless of which attribute above is picked, they must be passed into the account filter as `loginHint`.

#### Using `login_hint` claim

```javascript
const accountFilter = {
    loginHint: previouslyObtainedIdTokenClaims.login_hint;
};
request.account = myMSALObj.getAccountByFilter(accountFilter);
return await myMSALObj.acquireTokenSilent(request).catch(async (error) => {
    // Handle error
    return await myMSALObj.acquireTokenPopup(request);
});
```

#### Using `username``

```javascript
const accountUsername = userProfile.username;
const accountFilter = {

>     loginHint: accountUsername;

Reviewer 2: Based on the AccountFilter type above, could you also provide this as `username`?

Author: Yes you could, that was already enabled though. Considering it in login hint is mostly useful if you pick a login_hint out of whichever is available between the three and then stop caring what the source of the login_hint was (in case you lose context). So 10k lines above const loginHint was assigned to something, but you know for sure you can pass it into getAccount and it'll get compared to all the options available.

Reviewer 2: That's fair, if you know you have a username however, providing it as username will be slightly more performant. We should call out both so folks don't think they must provide it as loginHint

Author: I added some context/disambiguation.
